File: nodes/set-object-property/set-object-property.process.py
"""
Process function for the Set Object Property node.
Set or add a key-value pair to an object, returning a new object with the property set.
"""
from typing import Any, Dict

async def process(inputs: Dict[str, Any], settings: Dict[str, Any], config: Dict[str, Any], nodeConfig: Dict[str, Any]) -> Dict[str, Any]:
    object_input = inputs.get("object")
    key = inputs.get("key")
    value = inputs.get("value")
    overwrite = settings.get("overwrite", True)
    
    # Validate inputs - create empty object if input is null/undefined/invalid
    if not isinstance(object_input, dict):
        new_object = {}
        if key and isinstance(key, str):
            new_object[key] = value
        return {
            "object": new_object,
            "key": key or "",
            "value": value,
            "was_new": True
        }
    
    if not isinstance(key, str):
        return {
            "object": object_input,
            "key": key or "",
            "value": value,
            "was_new": False
        }
    
    # Check if property already exists
    was_new = key not in object_input
    
    # If overwrite is false and property exists, don't change it
    if not overwrite and not was_new:
        return {
            "object": object_input,
            "key": key,
            "value": object_input[key],
            "was_new": False
        }
    
    # Create new object with the property set
    new_object = object_input.copy()
    new_object[key] = value
    
    return {
        "object": new_object,
        "key": key,
        "value": value,
        "was_new": was_new
    }

File: sdks/nodejs/nodes/set-object-property/set-object-property.process.js
/**
 * Process function for the Set Object Property node.
 * Set or add a key-value pair to an object, returning a new object with the property set.
 */
export default async ({inputs, settings, config}) => {
  const { object, key, value } = inputs;
  const { overwrite } = settings;
  
  // Validate inputs - create empty object if input is null/undefined/invalid
  if (!object || typeof object !== 'object') {
    const newObject = {};
    if (key && typeof key === 'string') {
      newObject[key] = value;
    }
    return {
      object: newObject,
      key: key || '',
      value: value,
      was_new: true
    };
  }
  
  if (!key || typeof key !== 'string') {
    return {
      object: object,
      key: key || '',
      value: value,
      was_new: false
    };
  }
  
  // Check if property already exists
  const was_new = !(key in object);
  
  // If overwrite is false and property exists, don't change it
  if (!overwrite && !was_new) {
    return {
      object: object,
      key: key,
      value: object[key],
      was_new: false
    };
  }
  
  // Create new object with the property set
  const newObject = { ...object };
  newObject[key] = value;
  
  return {
    object: newObject,
    key: key,
    value: value,
    was_new: was_new
  };
};
